File: lib/emulation/core_router.dart
import 'dart:io';

class CoreRouter {
  // Maps file extensions to their ideal libretro core prefix
  static const Map<String, String> _extensionToCore = {
    // NES
    'nes': 'fceumm',
    // SNES
    'smc': 'snes9x',
    'sfc': 'snes9x',
    // Sega Genesis / Mega Drive
    'md': 'genesis_plus_gx',
    'gen': 'genesis_plus_gx',
    // Game Boy Advance
    'gba': 'mgba',
    // PlayStation 1
    'bin': 'pcsx_rearmed',
    'cue': 'pcsx_rearmed',
    'iso': 'pcsx_rearmed',
  };

  /// Returns a list of all ROM file extensions supported by the router.
  static List<String> getSupportedExtensions() {
    return _extensionToCore.keys.toList();
  }

  /// Resolves the correct Libretro core filename for a given ROM path.
  static String resolveCore(String romPath) {
    final ext = romPath.split('.').last.toLowerCase();
    
    final corePrefix = _extensionToCore[ext] ?? 'fceumm'; // Default to NES

    if (Platform.isAndroid) {
      return '${corePrefix}_libretro_android.so';
    } else if (Platform.isIOS) {
      return '${corePrefix}_libretro_ios.dylib';
    } else {
      // Fallback for desktop/other
      return '${corePrefix}_libretro.so';
    }
  }
}

File: android/app/src/main/java/dev/seven_cgpalabs/mojosnap/ui/CoreRouter.kt
package dev.seven_cgpalabs.mojosnap.ui

object CoreRouter {
    private val extensionToCore = mapOf(
        "nes" to "fceumm",
        "smc" to "snes9x",
        "sfc" to "snes9x",
        "md" to "genesis_plus_gx",
        "sms" to "genesis_plus_gx",
        "gg" to "genesis_plus_gx",
        "gba" to "mgba",
        "bin" to "pcsx_rearmed",
        "cue" to "pcsx_rearmed",
        "iso" to "pcsx_rearmed",
        "img" to "pcsx_rearmed",
        "exe" to "dosbox_pure",
        "bat" to "dosbox_pure",
        "com" to "dosbox_pure",
        "zip" to "dosbox_pure",
        "gb" to "gambatte",
        "gbc" to "gambatte"
    )

    fun getSupportedExtensions(): List<String> {
        return extensionToCore.keys.toList()
    }

    fun resolveCore(romPath: String): String {
        val ext = romPath.substringAfterLast('.').lowercase()
        val corePrefix = extensionToCore[ext] ?: "fceumm"
        return "${corePrefix}_libretro_android.so"
    }
}
